dvc/tree/gdrive.py

```diff
@@ -510,35 +511,71 @@ def exists(self, path_info, use_dvcignore=True):
+        if recursive:
+            for entry in self.walk_files(path_info, detail=detail):
+                if detail:
+                    yield entry
+                else:
+                    yield entry.path
+            return None
 
-        if prefix:
-            dir_ids = self._ids_cache["dirs"].get(prefix[:2])
-            if not dir_ids:
-                return
+        cached = path_info.path in self._ids_cache["dirs"]
+        if cached:
+            dir_ids = self._ids_cache["dirs"][path_info.path]
         else:
-            dir_ids = self._ids_cache["ids"]
-        parents_query = " or ".join(
-            f"'{dir_id}' in parents" for dir_id in dir_ids
-        )
-        query = f"({parents_query}) and trashed=false"
+            dir_ids = self._path_to_item_ids(path_info.path)
+
+        if not dir_ids:
+            return None
 
+        query = " or ".join(f"'{dir_id}' in parents" for dir_id in dir_ids)
+        query = f"({query}) and trashed=false"
+
+        root_path = path_info.path
         for item in self._gdrive_list(query):
-            parent_id = item["parents"][0]["id"]
-            yield posixpath.join(
-                self._ids_cache["ids"][parent_id], item["title"]
-            )
+            item_path = posixpath.join(root_path, item["title"])
+            if detail:
+                if item["mimeType"] == FOLDER_MIME_TYPE:
+                    yield {"type": "directory", "name": item_path}
+                else:
+                    yield {
+                        "type": "file",
+                        "name": item_path,
+                        "size": item["fileSize"],
+                        "checksum": item["md5Checksum"],
+                    }
+            else:
+                yield item_path
+
+        if not cached:
+            self._cache_path_id(root_path, min(dir_ids))
+
+    def walk(
+        self, path_info, prefix=None, detail=False
+    ):  # pylint: disable=unused-argument
+        stack = deque([path_info])
+        while stack:
+            path_info = stack.pop()
+            files, directories = [], []
+            for entry in self.ls(path_info, detail=True):
+                path = path_info.replace(path=entry["name"])
+                if entry["type"] == "file":
+                    if detail:
+                        files.append(entry)
+                    else:
+                        files.append(path)
+                else:
+                    directories.append(path)
+
+            yield path_info, directories, files
+            stack.extendleft(directories)
 
     def walk_files(self, path_info, **kwargs):
-        use_prefix = kwargs.pop("prefix", False)
-        if path_info == self.path_info or not use_prefix:
-            prefix = None
-        else:
-            prefix = path_info.path
-        for fname in self._list_paths(prefix=prefix, **kwargs):
-            yield path_info.replace(fname)
+        for _, _, file_infos in self.walk(path_info, **kwargs):
```

Comment:
do we replace the prefix logic completely everywhere in the `status` optimization logic?

Reply:
I think it should work effectively for all recursive calls with/without prefix

Reply:
yeah, this question I think is not about this local change, more about the upstream changes:

https://github.com/iterative/dvc/blob/00461bb5dff898f8add7fb2311105bb24294f001/dvc/cache/base.py#L762

do we change this logic (that utilizes prefix I think) to estimate the remote size?